packages/react-utilities/src/utils/properties.ts

```diff
@@ -380,59 +365,47 @@ export const imgProperties = toObjectMap(htmlElementProperties, [
- * props set. Note that all data- and aria- prefixed attributes will be allowed.
- * NOTE: getNativeProps should always be applied first when adding props to a react component. The
- * non-native props should be applied second. This will prevent getNativeProps from overriding your custom props.
- * For example, if props passed to getNativeProps has an onClick function and getNativeProps is added to
- * the component after an onClick function is added, then the getNativeProps onClick will override it.
+ * Gets native supported props for an HTML element provided the allowance set. Use one of the property
+ * sets defined (`divProperties`, `buttonPropertes`, etc) to filter out supported properties from a given
+ * props set. Note that all `data-` and `aria-` prefixed attributes will be allowed.
  *
- * @public
+ * NOTE: `getNativeProps` should always be applied first when adding props to a React component. The
+ * non-native props should be applied second. This will prevent `getNativeProps` from overriding your custom props.
+ * For example, if props passed to `getNativeProps` has an `onClick` function and `getNativeProps` is added to
+ * the component after an `onClick` function is added, then the `getNativeProps` `onClick` will override it.
  * @param props - The unfiltered input props
- * @param allowedPropsNames - The array or record of allowed prop names.
+ * @param allowedPropsNames - The set of allowed prop names.
  * @returns The filtered props
  */
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 export function getNativeProps<T extends Record<string, any>>(
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   props: Record<string, any>,
-  allowedPropNames: string[] | Record<string, number>,
+  allowedPropNames: Set<string>,
   excludedPropNames?: string[],
+  extraPropNames?: string[],
 ): T {
   // It'd be great to properly type this while allowing 'aria-` and 'data-' attributes like TypeScript does for
   // JSX attributes, but that ability is hardcoded into the TS compiler with no analog in TypeScript typings.
   // Then we'd be able to enforce props extends native props (including aria- and data- attributes), and then
   // return native props.
   // We should be able to do this once this PR is merged: https://github.com/microsoft/TypeScript/pull/26797
 
-  const isArray = Array.isArray(allowedPropNames);
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   const result: Record<string, any> = {};
   const keys = Object.keys(props);
 
   for (const key of keys) {
     const isNativeProp =
-      (!isArray && (allowedPropNames as Record<string, number>)[key]) ||
-      (isArray && (allowedPropNames as string[]).indexOf(key) >= 0) ||
-      key.indexOf('data-') === 0 ||
-      key.indexOf('aria-') === 0;
+      allowedPropNames.has(key) || extraPropNames?.includes(key) || key.startsWith('data-') || key.startsWith('aria-');
```

Comment:
@dzearing can correct me, but usage of objects was intentional was due performance as this gets called on every of every render of a component. I made a quick bench.

#### Chrome Headless 96.0.4664.93 (Windows 11)

```
- obj                                     935612481 ops/sec 
- arr (includes)                             106579 ops/sec 
- arr (indexOf)                                 353 ops/sec 
- set                                          5724 ops/sec 
```

#### Firefox 94.0 (Windows 11)

```
- obj                                     922216616 ops/sec 
- arr (includes)                                143 ops/sec 
- arr (indexOf)                                 167 ops/sec 
- set                                         58915 ops/sec
```

---

I used my setup with Rollup + Karma, https://github.com/layershifter/fela-benchmark

<details>
  <summary>Code</summary>


```tsx
const obj = {
  onAuxClick: 1,
  onCopy: 1,
  onCut: 1,
  onPaste: 1,
  onCompositionEnd: 1,
  onCompositionStart: 1,
  onCompositionUpdate: 1,
  onFocus: 1,
  onFocusCapture: 1,
  onBlur: 1,
  onBlurCapture: 1,
  onChange: 1,
  onInput: 1,
  onSubmit: 1,
  onLoad: 1,
  onError: 1,
  onKeyDown: 1,
  onKeyDownCapture: 1,
  onKeyPress: 1,
  onKeyUp: 1,
  onAbort: 1,
  onCanPlay: 1,
  onCanPlayThrough: 1,
  onDurationChange: 1,
  onEmptied: 1,
  onEncrypted: 1,
  onEnded: 1,
  onLoadedData: 1,
  onLoadedMetadata: 1,
  onLoadStart: 1,
  onPause: 1,
  onPlay: 1,
  onPlaying: 1,
  onProgress: 1,
  onRateChange: 1,
  onSeeked: 1,
  onSeeking: 1,
  onStalled: 1,
  onSuspend: 1,
  onTimeUpdate: 1,
  onVolumeChange: 1,
  onWaiting: 1,
  onClick: 1,
  onClickCapture: 1,
  onContextMenu: 1,
  onDoubleClick: 1,
  onDrag: 1,
  onDragEnd: 1,
  onDragEnter: 1,
  onDragExit: 1,
  onDragLeave: 1,
  onDragOver: 1,
  onDragStart: 1,
  onDrop: 1,
  onMouseDown: 1,
  onMouseDownCapture: 1,
  onMouseEnter: 1,
  onMouseLeave: 1,
  onMouseMove: 1,
  onMouseOut: 1,
  onMouseOver: 1,
  onMouseUp: 1,
  onMouseUpCapture: 1,
  onSelect: 1,
  onTouchCancel: 1,
  onTouchEnd: 1,
  onTouchMove: 1,
  onTouchStart: 1,
  onScroll: 1,
  onWheel: 1,
  onPointerCancel: 1,
  onPointerDown: 1,
  onPointerEnter: 1,
  onPointerLeave: 1,
  onPointerMove: 1,
  onPointerOut: 1,
  onPointerOver: 1,
  onPointerUp: 1,
  onGotPointerCapture: 1,
  onLostPointerCapture: 1,
};

const set = new Set([
  "onAuxClick",
  "onCopy",
  "onCut",
  "onPaste",
  "onCompositionEnd",
  "onCompositionStart",
  "onCompositionUpdate",
  "onFocus",
  "onFocusCapture",
  "onBlur",
  "onBlurCapture",
  "onChange",
  "onInput",
  "onSubmit",
  "onLoad",
  "onError",
  "onKeyDown",
  "onKeyDownCapture",
  "onKeyPress",
  "onKeyUp",
  "onAbort",
  "onCanPlay",
  "onCanPlayThrough",
  "onDurationChange",
  "onEmptied",
  "onEncrypted",
  "onEnded",
  "onLoadedData",
  "onLoadedMetadata",
  "onLoadStart",
  "onPause",
  "onPlay",
  "onPlaying",
  "onProgress",
  "onRateChange",
  "onSeeked",
  "onSeeking",
  "onStalled",
  "onSuspend",
  "onTimeUpdate",
  "onVolumeChange",
  "onWaiting",
  "onClick",
  "onClickCapture",
  "onContextMenu",
  "onDoubleClick",
  "onDrag",
  "onDragEnd",
  "onDragEnter",
  "onDragExit",
  "onDragLeave",
  "onDragOver",
  "onDragStart",
  "onDrop",
  "onMouseDown",
  "onMouseDownCapture",
  "onMouseEnter",
  "onMouseLeave",
  "onMouseMove",
  "onMouseOut",
  "onMouseOver",
  "onMouseUp",
  "onMouseUpCapture",
  "onSelect",
  "onTouchCancel",
  "onTouchEnd",
  "onTouchMove",
  "onTouchStart",
  "onScroll",
  "onWheel",
  "onPointerCancel",
  "onPointerDown",
  "onPointerEnter",
  "onPointerLeave",
  "onPointerMove",
  "onPointerOut",
  "onPointerOver",
  "onPointerUp",
  "onGotPointerCapture",
  "onLostPointerCapture",
]);

const arr = Object.keys(obj);

function lObj(prop) {
  return !!obj[prop];
}

function lArr(prop) {
  return arr.includes(prop);
}

function lArrIndexOf(prop) {
  return arr.indexOf(prop) !== -1;
}

function lSet(prop) {
  return set.has(prop);
}

// -----

let number = 1000;

function lookupObj(prop) {
  for (let i = 0; i <= number; i++) {
    lObj(prop);
  }
}

function lookupArr(prop) {
  for (let i = 0; i <= number; i++) {
    lArr(prop);
  }
}

function lookupArrIndexOf(prop) {
  for (let i = 0; i <= number; i++) {
    lArrIndexOf(prop);
  }
}

function lookupSet(prop) {
  for (let i = 0; i <= number; i++) {
    lSet(prop);
  }
}

suite.only("obj vs set vs arr", () => {
  benchmark(
    "obj",
    () => {
      lookupObj("onWheel");
      lookupObj("onPointerCancel");
      lookupObj("onPointerDown");
      lookupObj("onPointerEnter");
      lookupObj("onPointerLeave");
      lookupObj("onPointerMove");
      lookupObj("onPointerOut");
      lookupObj("onPointerOver");
      lookupObj("onPointerUp");
      lookupObj("onGotPointerCapture");
      lookupObj("onLostPointerCapture");

      lookupObj("onWheel1");
      lookupObj("onPointerCancel1");
      lookupObj("onPointerDown1");
      lookupObj("onPointerEnter1");
      lookupObj("onPointerLeave1");
      lookupObj("onPointerMove1");
      lookupObj("onPointerOut1");
      lookupObj("onPointerOver1");
      lookupObj("onPointerUp1");
      lookupObj("onGotPointerCapture1");
      lookupObj("onLostPointerCapture1");
    },
    { minSamples: 100 }
  );

  benchmark(
    "arr (includes)",
    () => {
      lookupArr("onWheel");
      lookupArr("onPointerCancel");
      lookupArr("onPointerDown");
      lookupArr("onPointerEnter");
      lookupArr("onPointerLeave");
      lookupArr("onPointerMove");
      lookupArr("onPointerOut");
      lookupArr("onPointerOver");
      lookupArr("onPointerUp");
      lookupArr("onGotPointerCapture");
      lookupArr("onLostPointerCapture");

      lookupArr("onWheel1");
      lookupArr("onPointerCancel1");
      lookupArr("onPointerDown1");
      lookupArr("onPointerEnter1");
      lookupArr("onPointerLeave1");
      lookupArr("onPointerMove1");
      lookupArr("onPointerOut1");
      lookupArr("onPointerOver1");
      lookupArr("onPointerUp1");
      lookupArr("onGotPointerCapture1");
      lookupArr("onLostPointerCapture1");
    },
    { minSamples: 100 }
  );

  benchmark(
    "arr (indexOf)",
    () => {
      lookupArrIndexOf("onWheel");
      lookupArrIndexOf("onPointerCancel");
      lookupArrIndexOf("onPointerDown");
      lookupArrIndexOf("onPointerEnter");
      lookupArrIndexOf("onPointerLeave");
      lookupArrIndexOf("onPointerMove");
      lookupArrIndexOf("onPointerOut");
      lookupArrIndexOf("onPointerOver");
      lookupArrIndexOf("onPointerUp");
      lookupArrIndexOf("onGotPointerCapture");
      lookupArrIndexOf("onLostPointerCapture");

      lookupArrIndexOf("onWheel1");
      lookupArrIndexOf("onPointerCancel1");
      lookupArrIndexOf("onPointerDown1");
      lookupArrIndexOf("onPointerEnter1");
      lookupArrIndexOf("onPointerLeave1");
      lookupArrIndexOf("onPointerMove1");
      lookupArrIndexOf("onPointerOut1");
      lookupArrIndexOf("onPointerOver1");
      lookupArrIndexOf("onPointerUp1");
      lookupArrIndexOf("onGotPointerCapture1");
      lookupArrIndexOf("onLostPointerCapture1");
    },
    { minSamples: 100 }
  );

  benchmark(
    "set",
    () => {
      lookupSet("onWheel");
      lookupSet("onPointerCancel");
      lookupSet("onPointerDown");
      lookupSet("onPointerEnter");
      lookupSet("onPointerLeave");
      lookupSet("onPointerMove");
      lookupSet("onPointerOut");
      lookupSet("onPointerOver");
      lookupSet("onPointerUp");
      lookupSet("onGotPointerCapture");
      lookupSet("onLostPointerCapture");

      lookupSet("onWheel1");
      lookupSet("onPointerCancel1");
      lookupSet("onPointerDown1");
      lookupSet("onPointerEnter1");
      lookupSet("onPointerLeave1");
      lookupSet("onPointerMove1");
      lookupSet("onPointerOut1");
      lookupSet("onPointerOver1");
      lookupSet("onPointerUp1");
      lookupSet("onGotPointerCapture1");
      lookupSet("onLostPointerCapture1");
    },
    { minSamples: 100 }
  );
});

```
</details>

Reply:
Just a shout out that even if perf requirements force us to move away from ES6 `Set`, the type changes in this PR still bring a lot of value to future proofing these native props

Reply:
That's unfortunate; I'm surprised Set isn't optimized better. I can revert that part of change due to perf impact (and add a comment).